docs/keps/20200619-federated-resource-status.md

```diff
@@ -0,0 +1,408 @@
+LabelRemovalTimedOut PropagationStatus = "LabelRemovalTimedOut"
+
+AggregateSuccess       AggregateReason = ""
+ComputePlacementFailed AggregateReason = "ComputePlacementFailed"
+NamespaceNotFederated  AggregateReason = "NamespaceNotFederated"
+
+PropagationConditionType ConditionType = "Propagation"
+```
+
+However the current federated resource properties don't help to track the status of the deployed resources in the kubefed clusters.
+
+The idea is to extend current GenericFederatedStatus with the Status of the resources:
+
+```go
+
+type GenericFederatedStatus struct {
+  	ObservedGeneration  int64                  `json:"observedGeneration,omitempty"`
+  	Conditions          []*GenericCondition    `json:"conditions,omitempty"`
+  	Clusters            []GenericClusterStatus `json:"clusters,omitempty"`
+}
+
+type GenericFederatedResource struct {
+	metav1.TypeMeta                `json:",inline"`
+	metav1.ObjectMeta              `json:"metadata,omitempty"`
+
+	Status *GenericFederatedStatus `json:"status,omitempty"`
+}
+
+```
+
+Nowadays `Conditions` hold the status of the federated actions (aka propagation status).
+In other words, it defines the conditions of the propagation status for a resource.
+
+```yaml
+- apiVersion: types.kubefed.io/v1beta1
+  kind: FederatedDeployment
+  metadata:
+    finalizers:
+    - kubefed.io/sync-controller
+    generation: 1
+    name: asystem
+    namespace: asystem
+    resourceVersion: "70174497"
+  spec:
+    placement:
+      clusters:
+      - name: cluster3
+      - name: cluster2
+      - name: cluster1
+    template:
+      metadata:
+        labels:
+          app: nginx
+      spec:
+        replicas: 3
+        selector:
+          matchLabels:
+            app: nginx
+        template:
+          metadata:
+            labels:
+              app: nginx
+          spec:
+            containers:
+            - image: nginx
+              name: nginx
+  status:
+    conditions:
+    - lastTransitionTime: "2020-05-25T19:47:59Z"
+      lastUpdateTime: "2020-05-25T19:47:59Z"
+      status: "True"
+      type: Propagation
+```
+
+`status.conditions` reports the latest status which defines the state of the propagation.
+Obviously it is not necessary to report all the clusters for which the propagation when
+successful.
+
+The intention in this proposal is to extend the current available `Conditions` to
+hold the status of the federated resources, e.g Ready, NotReady.
+
+The status of the federated resources should determine whether the resources satisfy
+a `Ready` condition, and otherwise report their error status.
+To do so, this property reports the status of the federated resources in their
+target clusters whenever a `ReadyCondition` is not satisfied.
+This condition would need to be identified per type or by the usage of an interface
+`IsReady` that determines this value per type of resource.
+By doing so, we ensure the `Conditions` property shows the status of only unhealthy
+resources.
```

Comment:
Maybe it's better to show status in all clusters, not only the unhealthy ones. e.g.
```
status:  
  clusters:
  - name: cluster1
    status："Ready"
  - name: cluster2
    status: "NotReady"
  conditions:
    ...
```

The `.status.clusters` should give cleary info.

Reply:
Imagine you have 100 clusters, that won't scale :/, that is why I understand they only saw the clusters for those whose propagation failed

Reply:
Yeah, I also thought about it. But, given users may use `spec.placement.clusterSelector` to specific clusters they want to deploy, It's better to clearly show the status, which means users can clearly know which clusters were matched.

There may be hundreds of items in `status.clusters`, it's a little bit not good for human reading, but I think it is sustainable.
Here is an example, that is `Node.Status.Images` which could contain a whole list of images in a specific node.

Reply:
I think having hundreds or thousands values in the status when a resource got propagated, it is hard to read/maintain :/. I'd like to hear why this was not done in the beginning by kubefed owners although I imagine this won't scale in these situations.

Reply:
The choice of storing only failure status compared to all is simply to reduce the load on etcd. We infact did play with an option to collect each resource status from each cluster as is into the federated resource using the controller [here](https://github.com/kubernetes-sigs/kubefed/blob/master/pkg/controller/status/controller.go). We did then implement the current propagation status mechanism to be able to:
- easily decipher the propagation failures.
- reduce load on etcd, which goes directly proportional to all resources in all clusters.

Currently all resources status is collected [only for services](https://github.com/kubernetes-sigs/kubefed/blob/master/pkg/apis/core/v1beta1/federatedtypeconfig_types.go#L218) if the service discovery feature is enabled. We could as well update this to make it a feature to users who understand the risk and be able to collect raw status side by side with the propagation status.

Reply:
> easily decipher the propagation failures.
   reduce load on etcd, which goes directly proportional to all resources in all clusters.

This makes a lot of sense to me. Honestly I would only like to see StatusReady or AnyFailed status for those resources that aren't. If I federate 1k resources showing them as Ready or propagated will be a killer in my opinion. 
Besides, I don't think anyone else requested to be able to report the success for all the federated resources yet.

I am trying to limit the scope of this KEP, so adding this feature to show the status of all the resources sounds to me like an additional feature or iteration. Can we agree on keeping the status reporting as done until now based on the initial approach ?

* easily decipher the propagation failures.
* reduce load on etcd, which goes directly proportional to all resources in all clusters.

Reply:
Sound good to me!